telemetry/prometheus.go

```diff
@@ -0,0 +1,109 @@
+package telemetry
+
+import (
+	"log"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type MetricType int
+
+const (
+	_ MetricType = iota
+	CounterType
+	GaugeType
+)
+
+type ToTelemetryRequest func(payload MetricPayload)
+
+type MetricPayload struct {
+	Labels map[string]string
+	Name   string
+	Value  float64
+}
+
+type Metric struct {
+	typ                MetricType
+	labels             map[string]string
+	toTelemetryRequest ToTelemetryRequest
+}
+
+type PrometheusMetrics struct {
+	metrics map[string]Metric
+}
+
+func NewPrometheusMetrics() *PrometheusMetrics {
+	return &PrometheusMetrics{
+		metrics: make(map[string]Metric),
+	}
+}
+
+func (pm *PrometheusMetrics) Register(name string, typ MetricType, labels prometheus.Labels, toTelemetryRequest ToTelemetryRequest) {
+	pm.metrics[name] = Metric{typ, labels, toTelemetryRequest}
+}
+
+func (pm *PrometheusMetrics) Snapshot() {
+	gatherer := prometheus.DefaultGatherer
+	metrics, err := gatherer.Gather()
+	if err != nil {
+		log.Fatalf("Failed to gather metrics: %v", err)
```

Comment:
Not sure if we should use `Fatalf`. If i remember this `panic`s, and IMO not being able to push metrics isn't an end of the world situation!